Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,405 @@
+  [MultisetList.v](https://github.com/fblanqui/color/blob/1.8.5/Util/Multiset/MultisetOrder.v).
+
+-/
+
+namespace Multiset
+
+variable {α : Type*}
+
+/-- The standard Dershowitz–Manna ordering: -/
+def DMLT [Preorder α] (M N : Multiset α) : Prop :=
+  ∃ (X Y Z : Multiset α),
+      Z ≠ ∅
+    ∧ M = X + Y
+    ∧ N = X + Z
+    ∧ (∀ y ∈ Y, ∃ z ∈ Z, y < z)
+
+/-- A special case of DMLT. The transitive closure of it is used to define
+    an equivalent (proved later) version of the ordering. -/
+def DMLT_singleton [LT α] (M N : Multiset α) : Prop :=
+  ∃ (X Y : Multiset α) (a : α) ,
+      (M = X + Y)
+    ∧ (N = X + {a})
+    ∧ (∀ y, y ∈ Y → y < a)
+
+open Relation
+
+/-- The transitive closure of DMLT_singleton and is equivalent to DMLT
+    (proved later). -/
+def TransLT [LT α] : Multiset α → Multiset α → Prop :=
+    TransGen DMLT_singleton
+
+/-- A special case of DMLT. -/
+theorem dmlt_of_DMLT_singleton [Preorder α] (M N : Multiset α)
+    (h : DMLT_singleton M N) : DMLT M N := by
+  rcases h with ⟨X, Y, a, M_def, N_def, ys_lt_a⟩
+  use X, Y, {a}, by simp, M_def, N_def
+  · simpa
+
+lemma DMLT_singleton_insert [DecidableEq α] [LT α] {a : α} {M N : Multiset α}
+    (h : DMLT_singleton N (a ::ₘ M)) :
+    ∃ M', N = (a ::ₘ M') ∧ (DMLT_singleton M' M) ∨ (N = M + M') ∧ (∀ x : α, x ∈ M' → x < a) := by
+  rcases h with ⟨X, Y, a0, h1, h0, h2⟩
+  by_cases hyp : a = a0
+  · exists Y; right; apply And.intro
+    · rw [h1, add_left_inj]
+      rw [add_comm, singleton_add] at h0
+      simp_all
+    · simp_all
+  · exists (Y + (M - {a0}))
+    left
+    constructor
+    · rw [h1]
+      have : X = (M - {a0} + {a}) := by
+        rw [add_comm, singleton_add] at *
+        ext b
+        simp only [ext, count_cons] at h0
+        by_cases h : b = a
+        · have := h0 b
+          simp_all only [ite_true, ite_false, add_zero, sub_singleton, count_cons_self, ne_eq,
+            not_false_eq_true, count_erase_of_ne]
+        · have := h0 b
+          simp [sub_singleton]
+          simp_all only [↓reduceIte, add_zero, ne_eq, not_false_eq_true, count_cons_of_ne]
+          split at this
+          next h_1 =>
+            simp_all only [count_erase_self, add_tsub_cancel_right]
+          next h_1 => simp_all only [add_zero, ne_eq, not_false_eq_true, count_erase_of_ne]
+      subst this
+      rw [add_comm]
+      nth_rewrite 2 [add_comm]
+      rw [singleton_add, add_cons]
+    · unfold DMLT_singleton
+      refine ⟨M - {a0}, Y, a0, ?_, ?_, h2⟩
+      · change Y + (M - {a0}) = (M - {a0}) + Y
+        rw [add_comm]
+      · change M = M - {a0} + {a0}
+        have this0: M = X + {a0} - {a} := by
+          rw [← h0, sub_singleton, erase_cons_head]
+        have a0M: a0 ∈ M := by
+          rw [this0, sub_singleton, mem_erase_of_ne]
+          rw [mem_add, mem_singleton]
+          · apply Or.inr
+            rfl
+          · exact fun h ↦ hyp (Eq.symm h)
+        rw [add_comm]
+        simp_all [singleton_add]
```

Comment:
Try to follow the line of sight rule. It says that the main branch of your code should be less indented
```suggestion
  exists (Y + (M - {a0}))
  left
  constructor
  · rw [h1]
    have : X = (M - {a0} + {a}) := by
      rw [add_comm, singleton_add] at *
      ext b
      simp only [ext, count_cons] at h0
      by_cases h : b = a
      · have := h0 b
        simp_all only [ite_true, ite_false, add_zero, sub_singleton, count_cons_self, ne_eq,
          not_false_eq_true, count_erase_of_ne]
      have := h0 b
      simp [sub_singleton]
      simp_all only [↓reduceIte, add_zero, ne_eq, not_false_eq_true, count_cons_of_ne]
      split at this
      next h_1 =>
        simp_all only [count_erase_self, add_tsub_cancel_right]
      next h_1 => simp_all only [add_zero, ne_eq, not_false_eq_true, count_erase_of_ne]
    subst this
    rw [add_comm]
    nth_rewrite 2 [add_comm]
    rw [singleton_add, add_cons]
  · unfold DMLT_singleton
    refine ⟨M - {a0}, Y, a0, ?_, ?_, h2⟩
    · change Y + (M - {a0}) = (M - {a0}) + Y
      rw [add_comm]
    · change M = M - {a0} + {a0}
      have this0 : M = X + {a0} - {a} := by
        rw [← h0, sub_singleton, erase_cons_head]
      have a0M : a0 ∈ M := by
        rw [this0, sub_singleton, mem_erase_of_ne]
        rw [mem_add, mem_singleton]
        · apply Or.inr
          rfl
        · exact fun h ↦ hyp (Eq.symm h)
      rw [add_comm]
      simp_all [singleton_add]
```